check that all accesses to registers have constant index expressions

skimo-openhub opened the pull request:
The index expressions of accesses to arrays of registers should
be constant since registers are not addressable.
Failing to use constant index expressions would result
in accesses to "local" memory, which is much slower than registers.
Some MatMulBias tests checked for particular patterns that should
not appear when the index expressions are constant, but they did
not check that they are actually constants.
Replace these checks by generic checks during emitCudaKernel
that explicitly check for constant index expressions and
that get applied to all accesses to registers.

Closes #517

ftynse commented on tc/core/polyhedral/cuda/codegen.cc:
```diff
@@ -396,10 +396,46 @@ struct LdgWrapper {
+  TC_CHECK(access);
+  for (int i = 1; i < access.get_n_arg(); ++i) {
+    auto arg = access.get_arg(i);
+    TC_CHECK(arg.as<isl::ast_expr_int>());
```
Let's have a description here saying something like "expected constant subscripts, got <..>".  Not sure if AST nodes can be simply converted to strings, no problem if not.

ftynse commented on tc/core/polyhedral/cuda/codegen.cc:
```diff
@@ -396,10 +396,46 @@ struct LdgWrapper {
+
+// Print an access to a(n array of) register(s), checking that
+// the index expressions are constant.
+template <typename AFF>
```
Why do we need a template here?  As far as I can see, there is only one caller of this function so there is no need to support different types.

ftynse commented on tc/core/polyhedral/cuda/codegen.cc:
```diff
@@ -396,10 +396,46 @@ struct LdgWrapper {
+// Print an access to global memory, wrapping the access in an "__ldg()"
+// call if the accessed tensor is known to be read-only.
+template <typename AFF>
+void emitGlobalAccess(AFF access, const CodegenStatementContext& context) {
```
Same question for the template: there are two call sites using the same argument type.

skimo-openhub commented:
On Thu, Jun 28, 2018 at 01:49:26AM -0700, ftynse wrote:
> +// Check that the given expression is an access with constant index expressions
> +void checkConstantAccess(isl::ast_expr expr) {
> +  auto op = expr.as<isl::ast_expr_op>();
> +  auto access = op.as<isl::ast_op_access>();
> +  TC_CHECK(access);
> +  for (int i = 1; i < access.get_n_arg(); ++i) {
> +    auto arg = access.get_arg(i);
> +    TC_CHECK(arg.as<isl::ast_expr_int>());
> 
> Let's have a description here saying something like "expected constant subscripts, got <..>".  Not sure if AST nodes can be simply converted to strings, no problem if not.

Added.

> +// Print an access to a(n array of) register(s), checking that
> +// the index expressions are constant.
> +template <typename AFF>
> 
> Why do we need a template here?  As far as I can see, there is only one caller of this function so there is no need to support different types.

*shrug*
I removed them.

skimo
